README.md
# Smart Enums
## Author: [Jop Middelkamp](https://github.com/jopmiddelkamp)

A package that can help you create 'smarter' enums that could be extended with some domain logic.

## Usage

```dart
final bestFramework = Framework.flutter;

print('${bestFramework == Framework.flutter}'); // Prints 'true'
print('${bestFramework == Framework.maui}'); // Prints 'false'

class Framework extends SmartEnum<Framework> {
  static const Framework flutter = Framework(
    name: 'Flutter',
    value: 1,
    isCrossPlatform: true,
    score: 10,
  );
  static const Framework maui = Framework(
    name: 'MAUI',
    value: 2,
    isCrossPlatform: true,
    score: 8,
  );
  static const Framework reactNative = Framework(
    name: 'React Native',
    value: 4,
    isCrossPlatform: true,
    score: 9,
  );

  const Framework({
    required String name,
    required int value,
    required this.isCrossPlatform,
    required this.score,
  }) : super(name, value);

  final bool isCrossPlatform;
  final int score;

  static List<Framework> get crossPlatformFrameworks => const [
        Framework.flutter,
        Framework.maui,
        Framework.reactNative,
      ].where((f) => f.isCrossPlatform).toList();
}
```

## Note

Inspired by [Steve Smith](https://github.com/ardalis)'s smart enum .NET [library](https://github.com/ardalis/SmartEnum).


CHANGELOG.md
## 1.0.4

* Updated the example app.

## 1.0.3

* Updated the Dart SDK version.

## 1.0.2

* Updated the description.

## 1.0.1

* Updated the read me.

## 1.0.0

* Initial release.

test/smart_enum_test.dart
import 'package:smart_enums/smart_enums.dart';
import 'package:test/test.dart';

void main() {
  group('SmarEnum', () {
    test('Does name gets stored', () {
      expect(Framework.flutter.name, equals('Flutter'));
      expect(Framework.maui.name, equals('MAUI'));
    });
    test('Does value gets stored', () {
      expect(Framework.flutter.value, equals(1));
      expect(Framework.maui.value, equals(2));
    });
    test('Does equality compareer compare by value', () {
      expect(Framework.flutter, equals(Framework.flutter));
      expect(Framework.maui, isNot(equals(Framework.flutter)));
    });
  });

  group('SmartEnumWithValue', () {
    test('Does name gets stored', () {
      expect(Padding.small.name, equals('small'));
      expect(Padding.regular.name, equals('regular'));
      expect(Padding.big.name, equals('regular'));
    });
    test('Does value gets stored', () {
      expect(Padding.small.value, equals(8));
      expect(Padding.regular.value, equals(16));
      expect(Padding.big.value, equals(24));
    });
    test('Does equality compareer compare by value', () {
      expect(Padding.regular, equals(Padding.regular));
      expect(Padding.regular, isNot(equals(Padding.big)));
    });
  });
}

class Padding extends SmartEnumWithValue<Padding, double> {
  static const Padding small = Padding(
    name: 'small',
    value: 8.0,
  );
  static const Padding regular = Padding(
    name: 'regular',
    value: 16.0,
  );
  static const Padding big = Padding(
    name: 'regular',
    value: 24.0,
  );

  const Padding({
    required String name,
    required double value,
  }) : super(name, value);
}

class Framework extends SmartEnum<Framework> {
  static const Framework flutter = Framework(
    name: 'Flutter',
    value: 1,
  );
  static const Framework maui = Framework(
    name: 'MAUI',
    value: 2,
  );
  static const Framework reactNative = Framework(
    name: 'MAUI',
    value: 4,
  );

  const Framework({
    required String name,
    required int value,
  }) : super(name, value);
}


lib/src/smart_enum.dart
/// A smarter enum that can be used to store additional data next to the value.
abstract class SmartEnum<TEnum> extends SmartEnumWithValue<TEnum, int> {
  /// Creates a new instance of the [SmartEnum] class.
  ///  - [name] The name of the enum.
  ///  - [value] The value of the enum.
  const SmartEnum(
    String name,
    int value,
  ) : super(name, value);
}

/// A smarter enum with a custom value that can be used to store additional data
/// as well.
abstract class SmartEnumWithValue<TEnum, TValue> {
  /// Creates a new instance of the [SmartEnumWithValue] class.
  ///  - [name] The name of the enum.
  ///  - [value] The value of the enum.
  const SmartEnumWithValue(
    this.name,
    this.value,
  );

  /// The name of the enum.
  final String name;

  /// The value of the enum.
  final TValue value;

  @override
  bool operator ==(Object other) =>
      other is SmartEnumWithValue<TEnum, TValue> && value == other.value;

  @override
  int get hashCode => value.hashCode;
}


example/main.dart
import 'package:smart_enums/smart_enums.dart';

void main() {
  final small = Padding.small;
  print('small: ${small == Padding.small}');
  print('regular: ${small == Padding.regular}');
  print('');
  final flutter = Framework.flutter;
  print('flutter: ${flutter == Framework.flutter}');
  print('maui: ${flutter == Framework.maui}');
  print('');
  print('Cross platform frameworks:');
  for (final framework in Framework.crossPlatformFrameworks) {
    print(' - ${framework.name}');
  }
}

class Padding extends SmartEnumWithValue<Padding, double> {
  static const Padding small = Padding(name: 'small', value: 8.0);
  static const Padding regular = Padding(name: 'regular', value: 16.0);
  static const Padding big = Padding(name: 'big', value: 24.0);

  const Padding({
    required String name,
    required double value,
  }) : super(name, value);
}

class Framework extends SmartEnum<Framework> {
  static const Framework flutter = Framework(
    name: 'Flutter',
    value: 1,
    isCrossPlatform: true,
    score: 10,
  );
  static const Framework maui = Framework(
    name: 'MAUI',
    value: 2,
    isCrossPlatform: true,
    score: 8,
  );
  static const Framework reactNative = Framework(
    name: 'React Native',
    value: 4,
    isCrossPlatform: true,
    score: 9,
  );

  const Framework({
    required String name,
    required int value,
    required this.isCrossPlatform,
    required this.score,
  }) : super(name, value);

  final bool isCrossPlatform;
  final int score;

  static List<Framework> get crossPlatformFrameworks => const [
        Framework.flutter,
        Framework.maui,
        Framework.reactNative,
      ].where((f) => f.isCrossPlatform).toList();
}
